imageCaptureProcessing maintain EXIF

melmathari opened the pull request:
Images taken in CommCare were losing their EXIF metadata, including critical GPS/location data, during processing. This created headaches for users who needed that juicy metadata intact, especially for location-based verification.

Resolves issue #2689 and #2743

## The Fix

We pimped the image capture and processing workflow to make EXIF data stick, no matter what. Here's what we did:

1. Injected EXIF-aware copying for raw image files.
2. Ensured EXIF data is kept intact during image scaling.
3. Made sure key EXIF tags stay in the game:

   - [x] GPS coordinates (latitude/longitude)
   - [x] GPS timestamps
   - [x] Image datetime
   - [x] Orientation metadata

## How We Pulled It Off

- [x] Created `copyFileWithExifData()` to keep EXIF data alive during file copying.
- [x] Introduced `scaleAndSaveImageWithExif()` for maintaining EXIF while scaling images.
- [x] Modified `ImageCaptureProcessing` to use our new EXIF-hugging methods.
- [x] Leveraged Android's `ExifInterface` API for bulletproof metadata handling.

## Test Drive

1. **Basic EXIF Check**
   - Snap a pic with GPS turned on.
   - Save the form.
   - Check that all EXIF data is intact using your favorite EXIF viewer.

2. **Scaled Image Test**
   - Take a high-res image.
   - Verify the EXIF data after it gets scaled down.

3. **Encrypted Image Test**
   - Enable media encryption in settings.
   - Capture a GPS-tagged image.
   - Confirm EXIF data survived the encryption.

4. **Edge Cases**
   - Use images with no EXIF data.
   - Test images from various sources (camera, gallery, etc.).
   - Try different formats (JPG, PNG).

## Tools to Verify

- Command line nerds: Use `exiftool` to peek at the EXIF data.

## Additional Notes

- Full backward compatibility with existing image operations.
- No database or form schema changes needed.
- Minimal performance impact, as EXIF operations run only on image files.

/claim #2689
/claim #2743

damagatchi commented:
Can one of the admins verify this patch?

melmathari commented:
/claim #2743

shubham1g5 commented on app/src/org/commcare/utils/FileUtil.java:
```diff
@@ -857,4 +858,72 @@ public static boolean isFileTooLargeToUpload(ContentResolver contentResolver, Ur
+        return scaled;
+    }
+
+    private static void copyExifData(ExifInterface source, ExifInterface dest) {
```
how about directly passing source and destination file paths to this method ?

shubham1g5 reviewed (changes requested):
Hey @melmathari. Your attempts looks great. Made some clarifying comments, but also wondering if it's possible to add unit tests for the new FileUtil methods you have added to copy the exif data. Thanks!

shubham1g5 commented:
@damagatchi ok to test

melmathari commented:
@shubham1g5 I've addressed the feedback. As for the unit tests, I'm not sure if I'll be able to complete them, i ran into several issues with     testImplementation project(path: ':commcare-core', configuration: 'testsAsJar')

shubham1g5 commented on app/src/org/commcare/utils/FileUtil.java:
```diff
@@ -121,7 +121,7 @@ public static boolean cleanFilePath(String fullPath, String extendedPath) {
     private static final String illegalChars = "'*','+'~|<> !?:./\\";
 
-    public static String SanitizeFileName(String input) {
+    public static String sanitizeFileName(String input) {
```
think you can ignore lint errors in the code you have not changed as method renames will need to be replicated while calling the method as well.

shubham1g5 replied:
Sorry if I wasn't clear but the file name will need to be reverted here, as otherwise the project fails to compile.

shubham1g5 commented on app/src/org/commcare/utils/FileUtil.java:
```diff
@@ -356,18 +356,17 @@ public static String getGlobalStringUri(String fileLocation) {
 
-    public static void checkReferenceURI(Resource r, String URI, Vector<MissingMediaException> problems) throws InvalidReferenceException {
-        Reference mRef = ReferenceManager.instance().DeriveReference(URI);
+    public static Reference checkReferenceUri(String uri) throws InvalidReferenceException {
```
this is causing the build to fail now due to method calls not having the corresponding change, you can ignore lint errors here.

shubham1g5 replied:
this will need to be reverted.

shubham1g5 reviewed (changes requested):
@melmathari It's fine to ignore the lint errors in the code you have not changed as the method refactoring will need to be replicated elsewhere in the calling code.

shubham1g5 commented:
>  As for the unit tests, I'm not sure if I'll be able to complete them, i ran into several issues with testImplementation project(path: ':commcare-core', configuration: 'testsAsJar')

Curious what kind of errors are you facing here and if you have followed the instructions [given here to run tests](https://github.com/dimagi/commcare-android?tab=readme-ov-file#run-unit-tests-from-the-command-line).

melmathari commented:
@shubham1g5 Thanks for providing the doc. Unfortunately, I won't be able to commit to the unit testing for this build due to my current schedule. We have tested this on our local implementation though.

shubham1g5 commented on app/lint.xml:
```diff
@@ -1,5 +1,9 @@
-    <issue id="LogNotTimber" severity="ignore" />
+    <issue id="LogNotTimber" severity="ignore">
+        <!-- Add specific paths to ignore -->
+        <ignore path="src/org/commcare/utils/FileUtil.java" />
+        <ignore path="src/org/commcare/activities/components/ImageCaptureProcessing.java" />
```
Sorry, I didn't mean to turn off lint checks for these files, but just to ignore any errors that you have not changed (lint errors doesn't block the merge on PR for us) .

melmathari commented:
@shubham1g5 Why did the developer bring a lint roller to the code review? Because we used SuppressLint to silence those pesky errors — nothing escapes our lint-taming prowess!

shubham1g5 commented on app/src/org/commcare/activities/components/ImageCaptureProcessing.java:
```diff
@@ -36,6 +36,7 @@ public class ImageCaptureProcessing {
      * @return A pair containing raw image and scaled imagePath. The first entry is the raw image
      * while the second one is path to scaled image.
      */
+    @SuppressLint("ExifInterface")
```
Sorry for the confusion but we don't want any `SuppressLint` annotations in the code even if the lint check breaks as we still want to know about all the lint errors in the code but just want to fix the ones that's been introduced by the new code in PR.

shubham1g5 commented on app/src/org/commcare/utils/FileUtil.java:
```diff
@@ -356,17 +357,16 @@ public static String getGlobalStringUri(String fileLocation) {
         try {
             if (!mRef.doesBinaryExist()) {
-                problems.addElement(new MissingMediaException(r, "Missing external media: " + mLocalReference, URI,
-                        MissingMediaException.MissingMediaExceptionType.FILE_NOT_FOUND));
+                throw new InvalidReferenceException("Missing external media: " + mLocalReference);
```
needs to be reverted as we want a `MissingMediaException`

shubham1g5 commented on app/src/org/commcare/utils/FileUtil.java:
```diff
@@ -356,17 +357,16 @@ public static String getGlobalStringUri(String fileLocation) {
+                throw new InvalidReferenceException("Missing external media: " + mLocalReference);
             }
         } catch (IOException e) {
-            problems.addElement(new MissingMediaException(r, "Problem reading external media: " + mLocalReference, URI,
```
needs to be reverted as well

shubham1g5 reviewed (changes requested):
@melmathari Indeed lol. Although what I have been trying to say is don't worry about the lint check and let it fail. No roller needed 🤣

shubham1g5 commented:
@melmathari We are seeing instrumentation test failures on this PR which looks related to media handling.


<img width="1098" alt="Screenshot 2024-11-26 at 11 26 23 AM" src="https://github.com/user-attachments/assets/9f828728-7fbc-4a74-9e20-5ac9b2dca546">


You should be able to run these tests locally to look into the failures.

melmathari commented:
@shubham1g5 Thanks, I will take some time to test this a bit more locally

shubham1g5 commented on app/src/org/commcare/utils/FileUtil.java:
```diff
@@ -857,4 +866,79 @@ public static boolean isFileTooLargeToUpload(ContentResolver contentResolver, Ur
+            try {
+                copyExifData(sourceFile.getAbsolutePath(), destFile.getAbsolutePath());
+            } catch (Exception e) {
+                Logger.log(LogTypes.TYPE_WARNING_NETWORK,
+                    "Failed to copy EXIF data: " + e.getMessage());
+            }
```
Seems like we are logging exceptions at 2 places with `copyExifData` being the other. Is it possible to just do it inside `copyExifData` and make it fail safe.

shubham1g5 reviewed (changes requested):
Curious if the error with tests was due to us trying to copy the exif data from non image files ?

melmathari commented:
> Curious if the error with tests was due to us trying to copy the exif data from non image files ?

Correct, this was the case.

melmathari commented:
@shubham1g5 I had to update my branch from the latest changes from the master branch.

melmathari commented:
@coderabbitai full review

melmathari commented:
@coderabbitai full review

melmathari commented:
@coderabbitai full review

ctsims commented:
@melmathari 

I wanted to reach out with my appreciation for this contribution! Image capture is key to some of our partner's critical workflows for everything from validating children's immunization histories to capturing photos of fields to support yield sizing. They'll appreciate this improved functionality. 

I wanted to reach out with a few small points to shepherd things through to wrap-up. 

1 - Don't worry too much about the Coderabbit feedback (of course feel free to follow if you find it helpful, of course). We are just launching the system on this particular repo, and haven't fully calibrated the inputs, so it's not mandatory for processing.

2 - Given the complexity of the testing harness and the straightforwardness of the code I'm comfortable with the structure of your testing approach for https://github.com/dimagi/commcare-android/issues/2689, but before closing out https://github.com/dimagi/commcare-android/issues/2743 I think it's important to have an artifact image in the PR to validate that we got the right EXIF property as set by devices and respected by browsers. Does your testing device have an EXIF output with orientation data set which you can provide to confirm that the image appears in a browser with the original faithful orientation (would be nice to have before/after, but not necessary). 

3 - Just a quick reminder from the [program terms](https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/2553577787/CommCare+Open+Source+Bounty+Program) that before we merge, we'll need [a CLA submission](https://docs.google.com/forms/d/1y44EykrBwLEPRtgQjg3epuOIzEzMc2Opn2UuJ_n0k_o/edit) from your end. 

Thanks!

shubham1g5 commented on app/src/org/commcare/activities/components/ImageCaptureProcessing.java:
```diff
@@ -66,17 +67,72 @@ private static Pair<File, String> moveAndScaleImage(File originalImage, boolean
+                    ExifInterface destExif = new ExifInterface(exifPath);
+                    
+                    // Copy all EXIF tags
+                    String[] tagsToPreserve = {
+                        ExifInterface.TAG_GPS_LATITUDE,
+                        ExifInterface.TAG_GPS_LATITUDE_REF,
+                        ExifInterface.TAG_GPS_LONGITUDE,
+                        ExifInterface.TAG_GPS_LONGITUDE_REF,
+                        ExifInterface.TAG_GPS_TIMESTAMP,
+                        ExifInterface.TAG_GPS_DATESTAMP,
+                        ExifInterface.TAG_GPS_ALTITUDE,
+                        ExifInterface.TAG_GPS_ALTITUDE_REF,
+                        ExifInterface.TAG_GPS_AREA_INFORMATION,
+                        ExifInterface.TAG_DATETIME,
+                        ExifInterface.TAG_DATETIME_DIGITIZED,
+                        ExifInterface.TAG_DATETIME_ORIGINAL,
+                        ExifInterface.TAG_OFFSET_TIME,
+                        ExifInterface.TAG_OFFSET_TIME_ORIGINAL,
+                        ExifInterface.TAG_OFFSET_TIME_DIGITIZED,
+                        ExifInterface.TAG_COPYRIGHT,
+                        ExifInterface.TAG_IMAGE_DESCRIPTION,
+                        ExifInterface.TAG_EXIF_VERSION,
+                        ExifInterface.TAG_ORIENTATION
+                    };
+                    
+                    for (String tag : tagsToPreserve) {
+                        String value = sourceExif.getAttribute(tag);
+                        if (value != null) {
+                            destExif.setAttribute(tag, value);
+                        }
+                    }
```
can we reuse code from `copyExifData` here ?

shubham1g5 commented on app/src/org/commcare/activities/components/ImageCaptureProcessing.java:
```diff
@@ -66,17 +67,72 @@ private static Pair<File, String> moveAndScaleImage(File originalImage, boolean
+                
+                // If we had EXIF data, store it in a companion file
+                if (sourceExif != null) {
+                    String exifPath = finalFilePath + ".exif";
```
I am a bit unsure how it's working. A few questions reg.  encrypted images - 

1. Curious why do we need to create a separate `.exif` file here to retain exif data ?

2. Before uploading images to the server, we decrypt them first by using a [EncryptedFileBody](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/network/EncryptedFileBody.java). Do we also need to upload the `.exif` file there to be able to retain exif data on server ?

3. In your testing, have you been able to test the exif retention for final image on server with encyrption enabled ?

kiran-tripathy commented:
@melmathari , I'm following up to check if you’re still planning to work on this. If not, we’d like to close it out to ensure it doesn’t block others who might be interested in taking this up. Let us know - thanks!

kiran-tripathy commented:
@melmathari Closing this out since there hasn't been any activity on this for a while, but please feel free to open a new one if you're interested in pursuing it.
